### PR TITLE
Fix for Callback state leak

### LINKED ISSUE #31: TaskCompletionSources are leaked when the callback is cancelled
The TaskCompletionSource is registered for later retrieval when a message arrives alongside with the message id in the `RequestResponseStateLookup`. When the request is cancelled the completion source remains in the lookup dictionary forever.

### PR DESCRIPTION
**Pull request by danielmarbach**
Fixes #31 

PR is based on #30 

* Pulls in the testable contexts to reproduce the leak
* Fixes TODO in tests for `When_sending_reply_to_the_request`
* Addresses memory leak
* Fixes bug with `TrySetCanceled`

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers please review. The first two commits will be gone as soon as #30 is merged

**Comment by danielmarbach** [comment]
Rebased

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers ready for review again

File: src/NServiceBus.Callbacks/IncomingContextExtensions.cs [hunk 1]
```diff
@@ -6,14 +6,13 @@ namespace NServiceBus
 
     static class IncomingContextExtensions
     {
-        public static CorrelationIdAndTaskCompletionSource GetCorrelationIdAndCompletionSource(this IIncomingContext context, IncomingMessage message, RequestResponseStateLookup lookup)
+        public static RequestResponseStateLookup.State? RemoveResponseStateBasedOnCorrelationIdIfAvailable(this IIncomingContext context, IncomingMessage message, RequestResponseStateLookup lookup)
```
Comment (timbussmann) [inline src/NServiceBus.Callbacks/IncomingContextExtensions.cs]: how about `TryRemoveResponseStateByCorrelationId`?

Comment (danielmarbach) [inline src/NServiceBus.Callbacks/IncomingContextExtensions.cs]: I wanted to avoid `TryRemove` since it implies returning a bool and having an out param

Comment (timbussmann) [inline src/NServiceBus.Callbacks/IncomingContextExtensions.cs]: I don't think TryXy implied an out parameter. The signature is clear enough imho.

**Comment by danielmarbach** [comment]
@timbussmann fixed

File: src/NServiceBus.Callbacks/ExtendableOptionsExtensions.cs [hunk 2]
```diff
@@ -8,19 +8,19 @@ static class ExtendableOptionsExtensions
...
             {
-                data.CancellationToken = cancellationToken;
+                state.CancellationToken = cancellationToken;
             }
             else
             {
-                data = new UpdateRequestResponseCorrelationTableBehavior.RequestResponseParameters
+                state = new RequestResponseStateLookup.State
                 {
                     CancellationToken = cancellationToken,
```
Comment (bording) [inline src/NServiceBus.Callbacks/ExtendableOptionsExtensions.cs]: The comma on this line got added in #30.

Comment (danielmarbach) [inline src/NServiceBus.Callbacks/ExtendableOptionsExtensions.cs]: AAARG

**Comment by danielmarbach** [comment]
@bording fixed

**Comment by andreasohlund** [comment]
All comments seems to be addressed and @danielmarbach walked me though this yesterday so I will merge this

**Pull request merged**